cEP-0015.md

```diff
@@ -0,0 +1,164 @@
+in order to harden the security.
+
+1. Give developers the right to invite newcomers to the organization.
+   This will involve changing the bot command to `corobo invite <@username>`.
+
+2. Remove the automatic invite due to `Hello World` message.
+
+3. Remove `invite me` bot command.
+
+4. Stop newcomers from getting assigned to more than one newcomer issue.
+
+5. Make all LabHub plugins require being a member of the organization. This
+   will be done by creating a `cmdfilter` to check if the user is member of
+   team or not.
+
+   ```
+   @cmdfilter
+   def check_membership(self, msg, cmd, args, dry_run)
+        usr = msg.frm.nik
+        teams = [self.TEAMS[self.GH_ORG_NAME + 'newcomers'],
+                 self.TEAMS[self.GH_ORG_NAME + 'developers'],
+                 self.TEAMS[self.GH_ORG_NAME + 'maintainers']
+                ]
+        for team in teams:
+            if team.is_member(usr):
+                return msg, cmd, args
+
+        return None, None, None
+   ```
+
+6. Add callback message to warn users who are spamming the room.
+
+7. Add bot command to ban a user from all rooms at once. The goal of
+   this feature is to prevent spamming as someone with many accounts
+   can't be stopped to join a room.
+
+## Tests
+
+The default TestBase provided by errbot is not enough for testing plugins like
+LabHub, which required intensive mocking.
+
+This will involve making changes upstream in
+[Errbot](https://github.com/errbotio/errbot/) and extend the existing testing
+infrastructure to implement better testing for plugins like LabHub.
+
+Errbot provides a pytest fixture `testbot` for testing purposes and to interact
+with the message queue. Extending the testbot and adding `inject_mock` method
+so that mocked objects can be injected into the plugin by passing a dictionary
+containing the `PluginName`, `field_name` and `Mocked Object`. The
+`inject_mock` method will inject a property on the plugin object and discard
+the assignments from the plugin itself.
+
+```
+def test_command(testbot):
+    mocks = {'PluginName': {'field_name1': fakeObj1,
+                                  'field_name2': fakeObj2}}
+              }
+    testbot.inject_mock(mocks)
+    assert 'something' in testbot.exec_command('!my_command')
+```
+
+## Configurability
+
+corobo has a potential to be used by other organizations for similar tasks like
+onboarding and automation. Currently, it is not configurable and many plugins
+are still very coala specific. Making it more configurable will allow other
+organizations to adapt corobo to cater their needs.
+
+Making existing coala specific plugins generic for other organizations and
+letting them configure the plugins as per their needs will ensure
+configurability.
+
+Errobot provides [plugin configuration](http://errbot.io/en/latest/user_guide/plugin_development/configuration.html)
+through the built-in `!config` command which can be used by other organizations
+to configure the plugins as per their needs.
```

Comment:
I created a [bunch of issues](https://github.com/coala/corobo/issues?utf8=%E2%9C%93&q=is%3Aopen+get_configuration_template+) about this, including some not mentioned below, but when I came to do it myself I found a problem is that we need configuration to be pre-loaded.

It isnt appropriate for a bot admin to need to run `!config` every time the bot starts up, for every plugin.
It needs to reload the previous saved state, or it needs to start using config from a static config file.

Maybe I missed something important.  See https://github.com/coala/corobo/issues/380#issuecomment-391575383 for a quick description of the problem and my workaround.

Reply:
@jayvdb I understand the problem but the configuration_template is there to avoid the need for admins to configure settings in some kind of configuration file, allowing configuration to happen directly through chat commands.

I guess configuring keys with empty `{TOKEN: ''}` values if they exist from the global config file will deal with this type of situation.

Reply:
@jayvdb  according to @gbin configurations are stored in ErrBot and last saved configuration are loaded on restart.

Reply:
there is still the problem of first time the bot is started.
the configuration of the corobo plugins needs to initialised somehow.

Reply:
Sure you here you have several choices: either a restore from a backup or you can insert values into storage from errbot's command line before startup. Errbot is not a stateless system but it can use existing external DBs as storage if you wish (SQL, redis, gcd etc.).

Reply:
We should choose a method, and document it here.

corobo is run in a docker, without any db, so a good choice would be a light-weight load method.

The reason why I wrote my patch above like that was so that the initial values of the plugin configuration was the same code as the plugin configuration code.

Reply:
I agree with @jayvdb on this one, reconfiguring again and again using the chat via a admin is not preferable. Also, it is worth noting that corobo is continuously deployed on each commit, hence, the usage of previous config stored in the db is not useful either. We need a way to preload the configuration.

Reply:
Can't we configure the plugins through [provisioning](http://errbot.io/en/latest/user_guide/provisioning.html#writing-values) before the bot is deployed on docker?

Reply:
try it and tell us if it works.
This is your project.

Reply:
We had a discussion on this point. 
It is possible to configure the plugins before deploying the bot.
We came to the conclusion that a python file containing the configurations will be mounted to the container from the host machine and we'll run a script from inside the container that will use the provisioning commands to configure the plugins from the configurations inside the python file before the bot is deployed.

Reply:
The configs you can set with `!plugin config` can be set with the command line under the key 'core' like the documentation shows. They are completely equivalent so I think it is relevant for your requirement for provisioning your plugins ahead of time.

Plugins have a place to validate their config is well formed by overriding `check_configuration` and will send to the bot admin a message explaining why a plugin could not start in case of failure.

Reply:
If they can be used together, that is great.

Below this cEP uses

```
errbot --storage-merge LabHub {'labhub': {'GH_ORG_NAME': 'coala'}}
```

Is that incorrect?  Should `LabHub` be `core`?

As mentioned above, I would want to see at least the same config variables in the `!config` example above and the `errbot --storage-merge` example below.

It would also be desirable to have a list of the config variables that this project will be implementing, so we can see and discuss them, and verify that the planned work is sufficient for the objective of the project.

Reply:
`errbot --storage-get core` will return configs of all the loaded plugins including LabHub,
 while `errbot --storage-get LabHub` will return config only for LabHub plugin.
Similar for `--storage-merge` and `--storage-set` flags,